cmd/zetae2e/local/bitcoin.go

```diff
@@ -5,71 +5,147 @@ import (
+	}
+
+	// create test routines
+	routineDeposit := createTestRoutine(runnerDeposit, depositTests)
+	routineWithdraw := createTestRoutine(runnerWithdraw, withdrawTests)
 
-		// funding the account
+	// start test routines
+	eg.Go(routineDeposit)
+	eg.Go(routineWithdraw)
+}
+
+// initRunnerDeposit initializes the runner for deposit tests
+func initRunnerDeposit(
+	conf config.Config,
+	deployerRunner *runner.E2ERunner,
+	verbose, initNetwork bool,
+) *runner.E2ERunner {
+	var (
+		name         = "btc_deposit"
+		account      = conf.AdditionalAccounts.UserBitcoin1
+		createWallet = true // deposit tests need Bitcoin node wallet to handle UTXOs
+	)
+
+	return initRunner(name, account, conf, deployerRunner, verbose, initNetwork, createWallet)
+}
+
+// initRunnerWithdraw initializes the runner for withdraw tests
+func initRunnerWithdraw(
+	conf config.Config,
+	deployerRunner *runner.E2ERunner,
+	verbose, initNetwork bool,
+) *runner.E2ERunner {
+	var (
+		name         = "btc_withdraw"
+		account      = conf.AdditionalAccounts.UserBitcoin2
+		createWallet = false // withdraw tests DON'T use Bitcoin node wallet
+	)
+
+	return initRunner(name, account, conf, deployerRunner, verbose, initNetwork, createWallet)
+}
+
+// initRunner initializes the runner for given test name and account
+func initRunner(
+	name string,
+	account config.Account,
+	conf config.Config,
+	deployerRunner *runner.E2ERunner,
+	verbose, initNetwork, createWallet bool,
+) *runner.E2ERunner {
+	// initialize runner for bitcoin test
+	runner, err := initTestRunner(
+		name,
+		conf,
+		deployerRunner,
+		account,
+		runner.NewLogger(verbose, color.FgYellow, name),
+	)
+	testutil.NoError(err)
+
+	// setup TSS address and setup deployer wallet
+	runner.SetupBitcoinAccounts(createWallet)
+
+	// initialize funds
+	if initNetwork {
+		// send some BTC block rewards to the deployer address
+		_, err = runner.GenerateToAddressIfLocalBitcoin(4, runner.BTCDeployerAddress)
+		require.NoError(runner, err)
+
+		// send ERC20 token on EVM
 		txERC20Send := deployerRunner.SendERC20OnEvm(account.EVMAddress(), 1000)
-		bitcoinRunner.WaitForTxReceiptOnEvm(txERC20Send)
+		runner.WaitForTxReceiptOnEvm(txERC20Send)
+
+		// deposit ETH and ERC20 tokens on ZetaChain
+		txEtherDeposit := runner.DepositEther()
+		txERC20Deposit := runner.DepositERC20()
 
-		// depositing the necessary tokens on ZetaChain
-		txEtherDeposit := bitcoinRunner.DepositEther()
-		txERC20Deposit := bitcoinRunner.DepositERC20()
+		runner.WaitForMinedCCTX(txEtherDeposit)
+		runner.WaitForMinedCCTX(txERC20Deposit)
+	}
 
-		bitcoinRunner.WaitForMinedCCTX(txEtherDeposit)
-		bitcoinRunner.WaitForMinedCCTX(txERC20Deposit)
+	return runner
+}
 
-		bitcoinRunner.SetupBitcoinAccount(initBitcoinNetwork)
-		bitcoinRunner.DepositBTC()
+// createTestRoutine creates a test routine for given test names
+// Note: due to the extensive block generation in Bitcoin localnet, block header test is run first
```

Comment:
> due to the extensive block generation in Bitcoin localnet
isn't there some optimization we could do on this side in the future?

Reply:
deleted https://github.com/zeta-chain/node/pull/3105/commits/03b7c9bc1c723a4924feb4490382a2a02cdb16d9